README.md
# amWiki 轻文库

![amWiki logo](https://github.com/YaoXuanZhi/amWikiForVSCode/blob/master/images/logo.png)  
amWiki 是一款基于 Javascript 脚本语言、依赖 Atom 编辑器、使用 Markdown 标记语法的轻量级开源 wiki 文库系统。  
amWiki 致力于让大家可以更简单、更便捷的建设个人和团队文库系统！  

[[view amWiki on Github](https://github.com/TevinLi/amWiki)]  
[![](https://img.shields.io/github/stars/TevinLi/amWiki.svg?style=social&label=Star)](https://github.com/TevinLi/amWiki "GitHub Stars") [![](https://img.shields.io/github/forks/TevinLi/amWiki.svg?style=social&label=Fork)](https://github.com/TevinLi/amWiki "GitHub Forks") [![](https://img.shields.io/github/issues-raw/TevinLi/amWiki.svg)](https://github.com/TevinLi/amWiki "GitHub Open Issues") [![](https://img.shields.io/github/issues-closed-raw/TevinLi/amWiki.svg)](https://github.com/TevinLi/amWiki "GitHub Closed Issues") [![](https://img.shields.io/github/contributors/TevinLi/amWiki.svg)](https://github.com/TevinLi/amWiki "GitHub Contributors")  
Apm:  
[![apm](https://img.shields.io/apm/v/amWiki.svg)](https://atom.io/packages/amWiki "Apm Version") [![apm](https://img.shields.io/apm/dm/amWiki.svg)](https://atom.io/packages/amWiki "Apm Downloads") [![apm](https://img.shields.io/apm/l/amWiki.svg)](https://atom.io/packages/amWiki "MIT License")

## 关于amWikiForVSCode
此项目是通过fork amWiki来的，由于amWiki是基于Atom的extension，因此在Visual Studio Code中是无法使用的，不过本人比较喜欢VS Code，于是着手将amWiki移植到VS Code中。幸亏Atom和VS Code都是基于electron二次开发的，所以两者的extension api有极高的相似度，因此移植工作还算进行得顺利。不过本人也是因为萌发了移植amWiki的念头之后，才开始学习VS Code extension开发的，因此有许多地方考虑不周，望大家见谅啦，如果发现有哪些地方的逻辑或者是语法使用不当的话，欢迎大家过来拍砖><。

[[view amWikiForVSCode on Github](https://github.com/YaoXuanZhi/amWikiForVSCode)]  
[![](https://img.shields.io/github/stars/YaoXuanZhi/amWikiForVSCode.svg?style=social&label=Star)](https://github.com/YaoXuanZhi/amWikiForVSCode "GitHub Stars") [![](https://img.shields.io/github/forks/YaoXuanZhi/amWikiForVSCode.svg?style=social&label=Fork)](https://github.com/YaoXuanZhi/amWikiForVSCode "GitHub Forks") [![](https://img.shields.io/github/issues-raw/YaoXuanZhi/amWikiForVSCode.svg)](https://github.com/YaoXuanZhi/amWikiForVSCode "GitHub Open Issues") [![](https://img.shields.io/github/issues-closed-raw/YaoXuanZhi/amWikiForVSCode.svg)](https://github.com/YaoXuanZhi/amWikiForVSCode "GitHub Closed Issues") [![](https://img.shields.io/github/contributors/YaoXuanZhi/amWikiForVSCode.svg)](https://github.com/YaoXuanZhi/amWikiForVSCode "GitHub Contributors")

## 在VSCode中安装amWiki插件
>在VSCode内置的应用商店上搜索、安装即可

![image](https://github.com/YaoXuanZhi/amWikiForVSCode/blob/master/res/Install_plugin.png)

## 在VSCode中的使用演示
- 创建新文库

> 先在一个文件夹下创建`config.json`文件，然后用`VSCode`打开此文件夹，此时在“`资源管理器`”上的`config.json`右击，执行右键菜单中的“基于config.json创建wiki”命令，即可在此文件夹下创建默认的amWiki文库，如下图所示：

![创建新文库](https://github.com/YaoXuanZhi/amWikiForVSCode/blob/master/res/Create_a_new_wiki.gif)

- 粘贴截图

>按下组合键`shift+ctrl+alt+v`或者是使用文档编辑器的右键菜单中的`粘帖截图到文档`命令，如下图所示：

![粘贴截图](https://github.com/YaoXuanZhi/amWikiForVSCode/blob/master/res/Paste_a_screenshot.gif)

- 插入页内目录

>使用文档编辑器的右键菜单中的`插入页内目录`命令，会把当前markdown文档中的h2、h3级作为页内目录抽取成文本，并将其粘贴到当前光标处，如下图所示：

![插入页内目录](https://github.com/YaoXuanZhi/amWikiForVSCode/blob/master/res/Insert_a_catalogue.gif)

- 导出项目为`github wiki`

>使用文档编辑器的右键菜单中的`导出项目为github wiki`命令，最终会在指定目录下生成符合github wiki目录结构的文件，如下图所示：

![导出项目为github wiki](https://github.com/YaoXuanZhi/amWikiForVSCode/blob/master/res/Export_to_disk_as_a_github_wiki.gif)

- 浏览当前页面

>按下组合键`f12`或者使用文档编辑器的右键菜单中的`浏览当前页面`命令，然后，会自动打开默认浏览器来浏览当前markdown文件对应的页面，如下图所示：

![在浏览器中浏览此页面](https://github.com/YaoXuanZhi/amWikiForVSCode/blob/master/res/Preview_the_current_page.gif)

- 快捷编辑ToDoList

>提供**Ctrl+Alt**+`[i|j|k|l]`组合键来快速编辑ToDoList，当然，你也可以自定义快捷键，默认ToDo样式为：`☐, ❑, ✘, ✔`，当然，你也可以将它改成SublimeText的PlainTasks风格><

![快捷编辑ToDoList](https://github.com/YaoXuanZhi/amWikiForVSCode/blob/master/res/Edit_todolist.gif)


CHANGELOG.md
# amWiki 版本日志

## v1.2.5
_2018-05-02_

- 添加todo快捷编辑功能
- 采用mathjax.js支持latex公式渲染

## [v1.2.1](https://github.com/TevinLi/amWiki/releases/tag/v1.2.1)
_2017-08-08_

- 改善 Mac 上点号开头的系统文件带来的诸多问题（例如导出报错等）([#32](https://github.com/TevinLi/amWiki/issues/32)、[#67](https://github.com/TevinLi/amWiki/issues/67))
- 允许在 config 上重新定义 library 路径 ([#47](https://github.com/TevinLi/amWiki/issues/47))
- 文章篇内目录显示优化 ([#33](https://github.com/TevinLi/amWiki/issues/33))
- 文章篇内目录序号识别，已有序号不再增加序号 ([#33](https://github.com/TevinLi/amWiki/issues/33))
- 基础 Markdown 语法扩充，实现文字飘红
- 统一官网和文档中心网址
- 修复几处显示细节

## [v1.2.0](https://github.com/TevinLi/amWiki/releases/tag/v1.2.0)
_2017-07-18_

- 新增本地数据挂载模式，可以直接双击 index 在浏览器中访问整个文库
    - Atom 平台和 Node-Npm 平台挂载数据操作绑定
    - 本地页面挂载数据生成与嵌入模块
    - 页面加载数据，页面打开时立即判断和读取挂载数据
    - 改进搜索，网络不可用时不再用 worker 而是直接搜索
    - disable 本地模式状态下的接口测试入口
    - 隐藏本地模式状态下的重建缓存按钮
- 基础 Markdown 语法扩充，支持 todo list 复选框 ([#56](https://github.com/TevinLi/amWiki/issues/56))
- 终端本地服务器命令新增可关闭 amWiki 索引页的可选参数
- 优化本地服务器容错性
- 搜索面板结果列表点击不再跳转页面
- 导航增加文档文件名转义，文档文件名允许使用语法符号`()[]'"`而不再冲突
- 对带`&`符号的文件名在URL地址上用特殊转义替代，不会再被作为参数截断 ([#64](https://github.com/TevinLi/amWiki/issues/64))
- 修复服务器文件系统为GBK编码时深目录无法读取的bug
- 调整大量注释、调整一些变量名，标记私有与公有

## [v1.1.3](https://github.com/TevinLi/amWiki/releases/tag/v1.1.3)
_2017-06-25_

- 增加大图预览模式第一期 ([#44](https://github.com/TevinLi/amWiki/issues/44))
- Markdown 基础语法扩充，支持定义图片大小与对齐方式 ([#45](https://github.com/TevinLi/amWiki/issues/45))
- 允许首页的文件名通过增加 `home-` 前缀来自定义
- 修复图片拖拽丢失扩展名的问题 ([#58](https://github.com/TevinLi/amWiki/issues/58))
- 更改 highlight.js 为完整版，支持 176 种语言语法高亮
- 修复测试面板切换文档时的一个显示问题

## [v1.1.2](https://github.com/TevinLi/amWiki/releases/tag/v1.1.2)
_2017-06-03_

- 允许在 config.json 上使用 imports 将额外的 css 和 js 文件嵌入页面 ([#43](https://github.com/TevinLi/amWiki/issues/43))
- 允许本地 server 多开，端口被占用时累加端口号 ([#46](https://github.com/TevinLi/amWiki/issues/46))
- 新增本地 server 索引页，本地根 url 不再 404
- 修复创建时取消创建操作的一个问题
- 调整一些文案

## [v1.1.1](https://github.com/TevinLi/amWiki/releases/tag/v1.1.1)
_2017-05-21_

- 紧急修复 win 平台 npm 创建报错的问题

## [v1.1.0](https://github.com/TevinLi/amWiki/releases/tag/v1.1.0)
_2017-05-20_

- 实现无限级导航 ([#5](https://github.com/TevinLi/amWiki/issues/5)、[#30](https://github.com/TevinLi/amWiki/pull/30))
    - 实现导航水平滚动条，调和水平垂直两滚动条冲突
    - 实现 PC、Mobi 端导航栏无限级梯度的层级显示
    - 实现目录无限级深度树形与列表数据读取
    - 实现无限级导航 Markdown 生成
    - 实现无限级重复的排序 ID 检查
    - 实现无限级图片粘贴、文件拖拽
- 优化工作端查找 library 文件夹
- 实现拖拽文件文件名汉字转拼音
- 修复导航初始化时展开的一个问题

---

### v1.0.6
_2017-05-05_

- 修复未开启测试模块的文库 web 端无内容显示的问题
- 修复导航未展开目录时显示不全只有一边的问题
- 修复上一版本 npm 失误的发布造成 Linux 平台命令无效的问题
- 移动端点击搜索按钮时关闭导航层 ([#40](https://github.com/TevinLi/amWiki/issues/40))

### v1.0.5
_2017-04-28_

- 导航筛选功能增强
    - 文件夹匹配
    - 筛选多个词
    - 忽略筛选词大小写
    - 筛选条件非主动删除不消失
    - 使用递归进行筛选，为无限级做准备 ([#5](https://github.com/TevinLi/amWiki/issues/5))
- 修复web端在移动端点击导航导航不自动收起的问题 ([#40](https://github.com/TevinLi/amWiki/issues/40))
- 修复web端测试面板在不关闭的条件下跳转文档后的一个显示问题
- 菜单栏增加 `从剪切板粘贴截图` 项，调和与其他插件快捷键冲突 ([#41](https://github.com/TevinLi/amWiki/issues/41))

### v1.0.4
_2017-04-22_

- 紧急修复命令行本地服务器不能启动的问题

### v1.0.3
_2017-04-22_

- 基础 Markdown 语法扩展，支持 footnote 脚注 ([#37](https://github.com/TevinLi/amWiki/issues/37))
- 修复 Windows 平台下 gitbash 执行命令结束报错的问题
- 修复 Linux 平台下 `/usr/bin/env` 解释器调用失败的问题
- 修复 Linux 平台下 create、version 报错的问题
- 释放快捷键绑定 Ctrl+D、Ctrl+H ([#38](https://github.com/TevinLi/amWiki/issues/38))
- 取消 readme 单篇长文介绍，仅保留安装方法，详细介绍引导至文档中心

### v1.0.2
_2017-04-04_

- 命令行启动服务器允许使用自定义端口参数
- 截图粘贴增强，QQ聊天、历史记录面板可复制单张图片直接粘入文档
- 修复命令行板块创建文库操作 $navigation.md 缺失的问题
- 修复本地 http 服务器 500 页面崩溃的问题
- 修复 web 端打开带 hash 地址链接时，hash 无效的问题
- 修复命令行显示版本号报错问题

### v1.0.1
_2017-03-30_

- 修复 npm 渠道安装失败的问题
- readme 添加 npm icon

### v1.0.0
_2017-03-30_

- [ _祝贺 amWiki 项目一周岁_ ]
- 构建基于 npm 的**命令行板块**，使用命令行可以同样创建维护文库
- 重构原工作端，分离为**构建板块**与**编辑器板块**
- 引入 ES6 异步编程，使用 Generator、yield、Thunk，将编辑器板块的 confirm 转为 confirm2，并与命令行板块的 confirm2 调用形式统一化
- 使用 ES6 异步编程，对构建板块部分回调方法进行重构，优先使用 return 与异步编程而不是 callback
- 使用 const/let 变量、解构赋值、箭头函数、Class 等 ES6 语法润色原工作端代码
- 增加命令行、终端界面信息打印着色、帮助菜单、版本显示
- 规范化部分公共方法注释

---

### v0.8.0
_2016-12-30_

- Web 端页面左右分栏滚动 ([#26](https://github.com/TevinLi/amWiki/issues/26))
- 工作端解除部分判断对 F12 造成的限制 ([#27](https://github.com/TevinLi/amWiki/issues/27))

---

### v0.7.7
_2016-12-19_  

- 全文库搜索，增加大小写忽略 ([#24](https://github.com/TevinLi/amWiki/issues/24))
- 修复导出后，在 github wiki 页内目录无法跳转的问题 ([#20](https://github.com/TevinLi/amWiki/issues/20#issuecomment-266927563))
- 修复导出时，html 格式的 img src 相对地址未转换问题 ([#23](https://github.com/TevinLi/amWiki/issues/23))
- 修复导出时，链接与 a 标签相对地址未转换问题 ([#23](https://github.com/TevinLi/amWiki/issues/23))

### v0.7.6
_2016-12-11_  

- 增加文库导出模块，可以将文库导出为 Github Wiki 项目 (**[#20](https://github.com/TevinLi/amWiki/issues/20)**)
- 允许 `.md` 文件文件名带空格 ([#19](https://github.com/TevinLi/amWiki/issues/19))
- Web 移动端点击页内目录以外的区域将折叠目录 ([#22](https://github.com/TevinLi/amWiki/issues/22))
- 修复某些情况下跳页面不返回顶部的问题 ([#21](https://github.com/TevinLi/amWiki/issues/21))

### v0.7.5
_2016-12-03_  
(none)

### v0.7.4
_2016-12-03_

- 增加 Web 端右侧目录悬浮窗 ([#15](https://github.com/TevinLi/amWiki/issues/15))
- 增加对 `[TOC]` 目录生成标签的支持 ([#15](https://github.com/TevinLi/amWiki/issues/15))
- 修复工作端抽取页内目录含单双引号、圆括号时无法正确解析的问题 ([#16](https://github.com/TevinLi/amWiki/pull/16))
- 修复工作端抽取页内目录无法跳转的问题 ([#15](https://github.com/TevinLi/amWiki/issues/15))
- 修复工作端抽取页内目录时内容为空报错的问题 ([#13](https://github.com/TevinLi/amWiki/issues/13), [#14](https://github.com/TevinLi/amWiki/pull/14))
- 修复 Web 端浏览器后退按钮报错问题
- Atom 菜单栏添加 “关于amWiki” 按钮

### v0.7.3
_2016-10-28_

- 修复部分atom下插件不能正常工作的问题

### v0.7.2
_2016-10-17_

- 新增搜索接口地址单独匹配并参与得分排序
- 新增搜索结果接口类文档专属标记
- 允许搜索框直接回车启动搜索 ([#12](https://github.com/TevinLi/amWiki/issues/12))
- Atom菜单微调整

### v0.7.1
_2016-10-10_

- 紧急修复 0.7.0 版本创建的 wiki 无法打开的问题

### v0.7.0
_2016-10-10_

- 新增全库搜索模块 ([#7](https://github.com/TevinLi/amWiki/issues/7))
    - 搜索worker子进程管理
    - 文档预处理实现
    - 关键词匹配得分排序机制设计
    - 搜索面板与基本操作
- 新增更新所有文档缓存
- 新增缓存统计
- 缓存模块其他优化
- 部分命名调整

---

### v0.6.5
_2016-10-04_  

- 新增一批基础说明文档
    - amWiki轻文库简介

lib/creatorplus.js
/**
 * 工作端 - 输出 web 静态资源模块
 * @author Tevin
 */

const fs = require('fs');
const co = require('./co');
const mngWiki = require('./manageWiki');
const mngFolder = require('./manageFolder');
const makeNav = require('./makeNavigation');
const vscode = require('vscode');

const creator = (function () {
    return {
        /**
         * 复制文件
         * @param {String} from
         * @param {String} to
         * @private
         */
        _copyFile: function (from, to) {
            fs.createReadStream(from).pipe(fs.createWriteStream(to));
        },
        /**
         * 创建 amWiki 必要的文件夹
         * @param {String} outputPath
         * @returns {Boolean}
         * @private
         */
        _createWikiFolder: function (outputPath) {
            if (!fs.existsSync(outputPath + 'amWiki/')) {
                fs.mkdirSync(outputPath + 'amWiki/', 0o777);
            }
            if (!fs.existsSync(outputPath + 'amWiki/js/')) {
                fs.mkdirSync(outputPath + 'amWiki/js/', 0o777);
            }
            if (!fs.existsSync(outputPath + 'amWiki/css/')) {
                fs.mkdirSync(outputPath + 'amWiki/css/', 0o777);
            }
            if (!fs.existsSync(outputPath + 'amWiki/images/')) {
                fs.mkdirSync(outputPath + 'amWiki/images/', 0o777);
            }
            if (!fs.existsSync(outputPath + 'library/')) {
                fs.mkdirSync(outputPath + 'library/', 0o777);
                if (!fs.existsSync(outputPath + 'library/001-学习amWiki/')) {
                    fs.mkdirSync(outputPath + 'library/001-学习amWiki/', 0o777);
                }
                if (!fs.existsSync(outputPath + 'library/001-学习amWiki/05-学习markdown/')) {
                    fs.mkdirSync(outputPath + 'library/001-学习amWiki/05-学习markdown/', 0o777);
                }
                if (!fs.existsSync(outputPath + 'library/002-文档示范/')) {
                    fs.mkdirSync(outputPath + 'library/002-文档示范/', 0o777);
                }
                return false;
            }
            return true;
        },
        /**
         * 创建着色色系
         *   (ps: 颜色变亮变暗不是 rgb 分别同时增加或减少一个值)
         * @param {String} color
         * @returns {{base: String, light: String, dark: String}} - 生成的浅、中、深三色
         * @private
         */
        _clacWikiColour: function (color) {
            const colours = {};
            //修复16进制下的1位数
            const lessThan = (str) => {
                if (str <= 9) {
                    str = '0' + str;
                } else {
                    const num = parseInt('0x' + str);
                    if (num >= 10 && num <= 15) {
                        str = '0' + str;
                    }
                }
                return str;
            };
            //颜色合法检查
            const atBinary16 = /^#([0-9a-fA-F]{3}){1,2}$/.test(color);
            const atRGB = /^[rR][gG][bB]([aA])?\(\d{1,3},\d{1,3},\d{1,3}(,.*?)?\)$/.test(color);
            const atList = /(blue|default)/.test(color);
            if (!atBinary16 && !atRGB && !atList) {
                vscode.window.showWarningMessage('请使用标准16进制颜色、标准RGB颜色、2种特定颜色名指定颜色！本次创建使用默认颜色');
                // alert('请使用标准16进制颜色、标准RGB颜色、2种特定颜色名指定颜色！本次创建使用默认颜色');
                return this._clacWikiColour('default');
            }
            //颜色名称转色值
            const colorList = {
                default: '#4296eb',
                blue: '#4296eb'
            };
            if (atList) {
                color = colorList[color];
            }
            //解析颜色为rgb
            const rgb = [];
            if (color.indexOf('#') === 0) {
                rgb[0] = parseInt('0x' + color.substr(1, 2));
                rgb[1] = parseInt('0x' + color.substr(3, 2));
                rgb[2] = parseInt('0x' + color.substr(5, 2));
            } else {
                const rgbStr = color.split('(')[1].split(')')[0].split(',');
                rgb[0] = parseInt(rgbStr[0]);
                rgb[1] = parseInt(rgbStr[1]);
                rgb[2] = parseInt(rgbStr[2]);
            }
            //生成主色base
            colours.base = '#' + [
                    lessThan(rgb[0].toString(16)),
                    lessThan(rgb[1].toString(16)),
                    lessThan(rgb[2].toString(16))
                ].join('');
            //颜色变化量，越小的值变化越大
            const increment = [
                Math.round((255 - rgb[0]) / 3),
                Math.round((255 - rgb[1]) / 3),
                Math.round((255 - rgb[2]) / 3)
            ];
            //生成亮色light
            colours.light = '#' + [
                    (rgb[0] + increment[0]).toString(16),
                    (rgb[1] + increment[1]).toString(16),
                    (rgb[2] + increment[2]).toString(16)
                ].join('');
            //生成暗色dark
            let larger = 0,
                largeIndex = -1;
            const largerFd = (i) => {
                if (increment[i] > larger) {
                    larger = increment[i];
                    largeIndex = i;
                }
            };
            largerFd(0);
            largerFd(1);
            largerFd(2);
            if (rgb[largeIndex] - larger < 0) {
                //修复加深颜色值为负的情况，最多减到0，其他三色等比例减小
                const reduce = (rgb[largeIndex] - larger) / larger;
                increment[0] += Math.round(increment[0] * reduce);
                increment[1] += Math.round(increment[1] * reduce);
                increment[2] += Math.round(increment[2] * reduce);
            }
            colours.dark = '#' + [
                    lessThan((rgb[0] - increment[0]).toString(16)),
                    lessThan((rgb[1] - increment[1]).toString(16)),
                    lessThan((rgb[2] - increment[2]).toString(16))
                ].join('');
            return colours;
        },
        /**
         * 配置检查
         * @param {String} configPath
         * @param {String} filesPath
         * @returns {Promise}
         * @private
         */
        _checkConfig: function (configPath, filesPath) {
            const that = this;
            return co(function*() {
                const options = {};
                //创建路径
                options.filesPath = filesPath;
                options.outputPath = configPath.split('config.json')[0].replace(/\\/g, '/');
                //读取配置
                let configStr = '';
                let noConfig = false;
                if (fs.existsSync(configPath)) {
                    configStr = fs.readFileSync(configPath, 'utf-8') || '';
                } else {
                    noConfig = true;
                }
                if (configStr.length === 0) {
                    vscode.window.showWarningMessage('没有读取到任何配置，继续创建么？', '继续创建').then(function (msg) {
                        if ('继续创建' === msg) {
                            config = "{}";
                            if (noConfig) {
                                fs.writeFileSync(configPath, configStr, 'utf-8');
                            }
                        } else {
                            return;
                        }
                    });
                    // if (yield confirm2('没有读取到任何配置，继续创建么？')) {
                    //     configStr = '{}';
                    //     if (noConfig) {
                    //         fs.writeFileSync(configPath, configStr, 'utf-8');
                    //     }
                    // } else {
                    //     return;
                    // }
                }
                //解析默认配置
                let parseOk = true,
                    config = null;
                try {
                    config = JSON.parse(configStr);
                } catch (e) {
                    // alert('配置解析失败，请检查您的 config.json 是否有严格按 Json 格式书写！\n错误消息：' + e.message);
                    vscode.window.showErrorMessage('配置解析失败，请检查您的 config.json 是否有严格按 Json 格式书写！\n错误消息：' + e.message);
                    parseOk = false;
                }
                if (!parseOk) {
                    return;
                }
                return {
                    options: options,
                    config: mngWiki.parseConfig(config)
                };
            }).catch((e) => {
                console.error(e);
            });
        },
        /**
         * 创建amWiki本地文件
         * @param {String} configPath - config.json文件的路径
         * @param {String} filesPath - 项目包files文件夹路径
         * @returns {Promise} 项目根目录
         * @public
         */
        create: function (configPath, filesPath) {
            const that = this;
            return co(function*() {
                const {options, config} = yield that._checkConfig(configPath, filesPath);
                //开始创建
                const files = mngFolder.readFolder(options.outputPath);
                if (files.length > 1) {
                     vscode.window.showWarningMessage('此处已有一些文件或文件夹，是否仍然在此创建amWiki？', '继续创建').then(function (msg) {
                        if ('继续创建' !== msg) {
                            return false;
                        }
                    });
                    // if (!(yield confirm2('此处已有一些文件或文件夹，是否仍然在此创建amWiki？'))) {
                    //     return false;
                    // }
                }
                //创建index.html
                let indexPage = fs.readFileSync(options.filesPath + 'amWiki.tpl', 'utf-8');
                indexPage = indexPage.replace(/\{\{name\}\}/g, config.name)
                    .replace('{{version}}', config.version)
                    .replace('{{logo}}', config.logo);
                //嵌入配置
                indexPage = indexPage.replace('{{config}}', 'AWConfig=' + JSON.stringify(config));
                //测试模块
                if (config.testing) {
                    const testingTpl = fs.readFileSync(options.filesPath + 'amWiki.testing.tpl', 'utf-8');
                    const testingScript = '<script src="amWiki/js/amWiki.testing.js"></script>';
                    indexPage = indexPage
                        .replace('{{amWiki.testing.tpl}}', testingTpl)
                        .replace('{{amWiki.testing.js}}', testingScript);
                } else {
                    indexPage = indexPage
                        .replace('{{amWiki.testing.tpl}}', '')
                        .replace('{{amWiki.testing.js}}', '');
                }
                //自定义 css、js 嵌入
                if (config.imports) {
                    let item,
                        cssList = '',
                        jsList = '';
                    for (item of config.imports.css) {
                        cssList = '<link rel="stylesheet" type="text/css" href="' + item + '" />'
                    }
                    for (item of config.imports.js) {
                        jsList = '<script type="text/javascript" src="' + item + '"></script>'
                    }
                    indexPage = indexPage
                        .replace('{{custom.css}}', cssList)
                        .replace('{{custom.js}}', jsList);
                } else {
                    indexPage = indexPage
                        .replace('{{custom.css}}', '')
                        .replace('{{custom.js}}', '');
                }
                fs.writeFileSync(options.outputPath + 'index.html', indexPage, 'utf-8');
                //创建文件夹
                const hasLibrary = that._createWikiFolder(options.outputPath);
                //创建amWiki.css
                let wikiCss = fs.readFileSync(options.filesPath + 'amWiki.css', 'utf-8');
                if (config.testing) {
                    wikiCss += fs.readFileSync(options.filesPath + 'amWiki.testing.css', 'utf-8');
                }
                wikiCss += fs.readFileSync(options.filesPath + 'amWiki.print.css', 'utf-8');
                wikiCss += fs.readFileSync(options.filesPath + 'amWiki.search.css', 'utf-8');
                let colors = that._clacWikiColour(config.colour);
                wikiCss = wikiCss
                    .replace(/@colour-base/g, colors.base)
                    .replace(/@colour-light/g, colors.light)
                    .replace(/@colour-dark/g, colors.dark);
                fs.writeFileSync(options.outputPath + 'amWiki/css/amWiki.css', wikiCss, 'utf-8');
                //拷贝页面资源
                const fileList = [
                    ['markdownbody.github.css', 'amWiki/css/markdownbody.github.css'],
                    ['lhjs.github-gist.css', 'amWiki/css/lhjs.github-gist.css'],
                    ['gbk.js', 'amWiki/js/gbk.js'],
                    ['pinyin.js', 'amWiki/js/pinyin.js'],
                    ['flowchart.min.js', 'amWiki/js/flowchart.min.js'],
                    ['raphael-min.js', 'amWiki/js/raphael-min.js'],
                    ['jquery-compat-3.1.0.min.js', 'amWiki/js/jquery-compat-3.1.0.min.js'],
                    ['marked.min.js', 'amWiki/js/marked.min.js'],
                    ['highlight.pack.js', 'amWiki/js/highlight.pack.js'],
                    ['amWiki.js', 'amWiki/js/amWiki.js'],
                    ['amWiki.scrollbar.js', 'amWiki/js/amWiki.scrollbar.js'],
                    ['amWiki.imgsView.js', 'amWiki/js/amWiki.imgsView.js'],
                    ['amWiki.testing.js', 'amWiki/js/amWiki.testing.js'],
                    ['amWiki.tools.js', 'amWiki/js/amWiki.tools.js'],
                    ['amWiki.storage.js', 'amWiki/js/amWiki.storage.js'],
                    ['amWiki.docs.js', 'amWiki/js/amWiki.docs.js'],
                    ['amWiki.search.js', 'amWiki/js/amWiki.search.js'],
                    ['amWiki.search.worker.js', 'amWiki/js/amWiki.search.worker.js'],
                    ['icons.svg', 'amWiki/images/icons.svg'],
                    ['logo.png', 'amWiki/images/logo.png'],
                    ['menubar_bg.png', 'amWiki/images/menubar_bg.png']
                ];
                for (let file of fileList) {
                    that._copyFile(options.filesPath + file[0], options.outputPath + file[1]);
                }
                //如果没有library则复制一套默认文档
                if (!hasLibrary) {
                    //首页文档
                    const home = fs.readFileSync(options.filesPath + 'doc.home.md', 'utf-8').replace('{{name}}', config.name);
                    fs.writeFileSync(options.outputPath + 'library/home-首页.md', home, 'utf-8');
                    //其他页面文档
                    let fileList2 = [
                        ['doc.amwiki-introduce.md', 'library/001-学习amWiki/01-amWiki轻文库简介.md'],
                        ['doc.amwiki-mind-map.md', 'library/001-学习amWiki/02-amWiki功能导图.md'],
                        ['doc.amwiki-new.md', 'library/001-学习amWiki/03-如何开始一个新amWiki轻文库.md'],
                        ['doc.amwiki-edit.md', 'library/001-学习amWiki/04-如何编辑amWiki轻文库.md'],
                        ['doc.amwiki-testing.md', 'library/001-学习amWiki/06-使用测试模块测试接口.md'],
                        ['doc.amwiki-test-cross.md', 'library/001-学习amWiki/07-amWiki转接到任意域名进行接口测试.md'],
                        ['doc.md-start.md', 'library/001-学习amWiki/05-学习markdown/01-Markdown快速开始.md'],
                        ['doc.md-high-lighting.md', 'library/001-学习amWiki/05-学习markdown/02-amWiki与语法高亮.md'],
                        ['doc.md-flow-chart.md', 'library/001-学习amWiki/05-学习markdown/03-amWiki与流程图.md'],
                        ['doc.md-atom.md', 'library/001-学习amWiki/05-学习markdown/05-Atom对Markdown的原生支持.md'],
                        ['doc.demo-api.md', 'library/002-文档示范/001-通用API接口文档示例.md'],
                        ['doc.demo-long-article.md', 'library/002-文档示范/002-超长文档页内目录示例.md']
                    ];
                    for (let file of fileList2) {
                        that._copyFile(options.filesPath + file[0], options.outputPath + file[1]);
                    }
                }
                return options.outputPath;
            }).catch((e) => {
                console.error(e);
            });
        }
    };
})();

module.exports = creator;

